Remove old code paths - all feature flags (FF_*)

jzbahrai opened the pull request:
# Summary | Résumé

Remove all code paths related to the below 4 configs being set to False:
1.     FF_NOTIFICATION_CELERY_PERSISTENCE = env.bool("FF_NOTIFICATION_CELERY_PERSISTENCE", True)
2.     FF_BATCH_INSERTION = env.bool("FF_BATCH_INSERTION", True)
3.     FF_REDIS_BATCH_SAVING = env.bool("FF_REDIS_BATCH_SAVING", True)
4.     FF_PRIORITY_LANES = env.bool("FF_PRIORITY_LANES", True)

The above 4 are true in prod.
This PR removes all extra code paths that occur when the above configs are `False`.

Tested by running unit test manually. Will further tests when on staging.

sastels commented:
Should we also delete `email_queue` and `sms_queue`?
https://github.com/cds-snc/notification-api/pull/1562/files#diff-9cec7b11237bc29d77a439e81c9b7acfac003d8e8855731eb6bc130b5a8ce602L51-L52

sastels commented:
Can also delete `beat_inbox_email()` and `beat_inbox_sms()`?
https://github.com/cds-snc/notification-api/pull/1562/files#diff-9cec7b11237bc29d77a439e81c9b7acfac003d8e8855731eb6bc130b5a8ce602L51-L52

sastels commented on app/celery/tasks.py:
```diff
@@ -1062,25 +1045,23 @@ def _acknowledge_notification(notification_type: Any, template: Any, receipt: UU
+        elif template.process_type == BULK:
+            email_bulk.acknowledge(receipt)
         try:
             email_queue.acknowledge(receipt)
```
Also this one

sastels commented on app/config.py:
```diff
@@ -499,10 +499,10 @@ class Config(object):
+    FF_NOTIFICATION_CELERY_PERSISTENCE = env.bool("FF_NOTIFICATION_CELERY_PERSISTENCE", True)
+    FF_BATCH_INSERTION = env.bool("FF_BATCH_INSERTION", True)
+    FF_REDIS_BATCH_SAVING = env.bool("FF_REDIS_BATCH_SAVING", True)
+    FF_PRIORITY_LANES = env.bool("FF_PRIORITY_LANES", True)
```
Do we still need these FF?

sastels commented on app/v2/notifications/post_notifications.py:
```diff
@@ -313,6 +311,8 @@ def process_sms_or_email_notification(*, form, notification_type, api_key, templ
         "client_reference": form.get("reference", None),
     }
 
+    current_app.logger.info(f"Notification id {notification['id']} sent to buffer queue.")
```
was this just for debugging or intentionally added?

sastels commented on pytest.ini:
```diff
@@ -2,8 +2,8 @@
-    FF_BATCH_INSERTION=False
-    FF_REDIS_BATCH_SAVING=False
+    FF_BATCH_INSERTION=True
+    FF_REDIS_BATCH_SAVING=True
```
🔥

sastels commented on tests/app/aws/test_metrics.py:
```diff
@@ -61,25 +59,14 @@ def test_put_batch_metric_multiple_items(self, mocker, metrics_logger_mock):
         metrics_logger_mock.put_metric.assert_called_with("batch_saving_published", 20, "Count")
 
     def test_put_batch_saving_in_flight_metric_FF_PRIORITY_LANES_true(self, mocker, metrics_logger_mock):
```
Can rename without the `_FF_PRIORITY_LANES_true` 🎉

sastels commented on tests/app/celery/test_scheduled_tasks.py:
```diff
@@ -610,19 +604,7 @@ def test_beat_inbox_email_priority(self, notify_api, mocker):
-        tasks.sms_queue.expire_inflights.assert_called_once()
-        tasks.email_queue.expire_inflights.assert_called_once()
-
     def test_recover_expired_notifications_priority_lanes(self, mocker, notify_api):
```
rename this one too

sastels commented on tests/app/celery/test_tasks.py:
```diff
@@ -519,39 +499,13 @@ def test_process_smss_job_metric(self, sample_template_with_placeholders, mocker
-        mocker.patch.object(Config, "FF_PRIORITY_LANES", False)
         mocker.patch("app.celery.tasks.s3.get_job_from_s3", return_value=load_example_csv("sms"))
-        mocker.patch("app.celery.tasks.save_sms.apply_async")
+        mocker.patch("app.celery.tasks.save_smss.apply_async")
```
Interesting! so before we were testing something like Priority Lanes on but batch saving off? That's messed up.

sastels commented on tests/app/celery/test_tasks.py:
```diff
@@ -824,54 +778,22 @@ def test_should_process_email_job_with_sender_id(self, sample_email_template, mo
         )
 
-    @freeze_time("2016-01-01 11:09:00.061258")
-    def test_should_process_letter_job(self, sample_letter_job, mocker):
```
I'm all for deleting letter code, by why this one?

jzbahrai replied:
actually the code path doesnt make sense anymore...

jzbahrai replied:
im just going to skip this test as this code path isn't around

sastels commented on tests/app/v2/notifications/test_post_notifications.py:
```diff
@@ -200,21 +112,13 @@ def test_post_sms_notification_uses_inbound_number_as_sender(notify_api, client,
-    assert resp_json["content"]["from_number"] == "1"
-    assert notifications[0].reply_to_text == "1"
-    mocked.assert_called_once_with([str(notification_id)], queue="send-sms-tasks")
+    assert mocked.called
```
I think we need to test that the mock gets the correct argument.

jzbahrai replied:
done

sastels commented on tests/app/v2/notifications/test_post_notifications.py:
```diff
@@ -200,21 +112,13 @@ def test_post_sms_notification_uses_inbound_number_as_sender(notify_api, client,
+    assert mocked.called
 
 
 def test_post_sms_notification_uses_inbound_number_reply_to_as_sender(notify_api, client, notify_db_session, mocker):
```
I'm confused as to what the difference is between this test and the previous... but there must be one since the first test previously used `deliver_sms` while this one used `deliver_throttled_sms`.... does the `inbound_number` make that difference? Since now we just throw things into redis here, do we need to move this test somewhere else?

sastels commented on tests/app/v2/notifications/test_post_notifications.py:
```diff
@@ -516,24 +371,6 @@ def test_post_email_notification_returns_201_with_celery_persistence(
-        )
-        in resp_json["template"]["uri"]
-    )
-    assert not resp_json["scheduled_for"]
```
do we have to delete all this? Seems that we should delete (or move) the other `test_post_email_notification_returns_201()` and keep this one roughly as it is (ie the post to /email doesn't actually put it into the database, but it does return a bunch of stuff like this to the user. Possibly this should be called something like `test_post_email_notification_returns_notification_information()`.

jzbahrai replied:
skipped

sastels commented on tests/app/v2/notifications/test_post_notifications.py:
```diff
@@ -881,10 +705,7 @@ def test_post_sms_notification_returns_201_if_allowed_to_send_int_sms_with_celer
 
 
 def test_post_sms_should_persist_supplied_sms_number(notify_api, client, sample_template_with_placeholders, mocker):
```
I think that the idea here is to test that the correct phone number is being put into the `to`. We could test the arguments to the `publish` mock.

sastels commented on tests/app/v2/notifications/test_post_notifications.py:
```diff
@@ -1156,10 +962,6 @@ def test_post_notification_with_document_upload(
 
-    notification = Notification.query.one()
-    assert notification.status == NOTIFICATION_CREATED
-    assert notification.personalisation == {"document": document_response}
```
hrm, this wouldn't be done before we put it into redis, but we should move this test to somewhere else.

jzbahrai replied:
skipped

jzbahrai commented on tests/app/v2/notifications/test_post_notifications.py:
```diff
@@ -1045,16 +857,11 @@ def test_post_email_notification_with_valid_reply_to_id_returns_201(notify_api,
     resp_json = json.loads(response.get_data(as_text=True))
     assert validate(resp_json, post_email_response) == resp_json
-    notification = Notification.query.first()
-    assert notification.reply_to_text == "test@test.com"
```
skipped

jzbahrai commented on tests/app/v2/notifications/test_post_notifications.py:
```diff
@@ -350,9 +234,6 @@ def test_notification_reply_to_text_is_original_value_if_sender_is_changed_after
     assert response.status_code == 201
-    notifications = Notification.query.all()
-    assert len(notifications) == 1
-    assert notifications[0].reply_to_text == "123456"
```
yes this test will need to be updated

jzbahrai commented:
@andrewleith - if you see the latest commit [2dc241f](https://github.com/cds-snc/notification-api/pull/1562/commits/2dc241f8bec58ce17b5a5596bfbc042587508e07) it addresses your comment. the code path save_sms, save_email and persist_notifcation are not used anymore. We need to move/ update these tests to address the newest code paths but we wont do this in this PR but another one.

jzbahrai commented:
@sastels I added in mocks where I could, but there is actual functionality tests we will still have to tackle. that can be done in another PR.

sastels reviewed (approved):
LGTM! @jimleroyer @andrewleith any thoughts?

andrewleith commented:
Yes, @jzbahrai has addressed my concerns, good to go.
